src/NUnitFramework/framework/Constraints/EmptyConstraint.cs

```diff
@@ -51,18 +51,20 @@ public override string Description
         public override ConstraintResult ApplyTo<TActual>(TActual actual)
         {
             // NOTE: actual is string will fail for a null typed as string
-            Type actualType = typeof(TActual);
+            Type actualType = actual?.GetType() ?? typeof(TActual);
             
             if (actualType == typeof(string))
                 realConstraint = new EmptyStringConstraint();
             else if (actual is Guid || actualType == typeof(Guid?))
                 realConstraint = new EmptyGuidConstraint();
-            else if (actual == null)
-                throw new System.ArgumentException($"The actual value must be a string, Guid, non-null IEnumerable or DirectoryInfo. The value passed was of type {actualType}.", nameof(actual));
             else if (actual is System.IO.DirectoryInfo)
                 realConstraint = new EmptyDirectoryConstraint();
-            else
+            else if (actual is System.Collections.IEnumerable)
                 realConstraint = new EmptyCollectionConstraint();
+            else if (CountZeroConstraint.HasCountProperty(actualType))
+                realConstraint = new CountZeroConstraint();
+            else
+                throw new ArgumentException($"The actual value must be a string, Guid, have a Count property, non-null IEnumerable or DirectoryInfo. The value passed was of type {actualType}.", nameof(actual));
```

Comment:
Thanks for tackling this @manfred-brands . I recently hit this myself with `NameValueCollection` in a netfx project.

If I'm reading the diff right, this should better advise on future edge cases where a non-null yet incompatible type is passed in as it should avoid a cryptic NRE.

Reply:
Yes, previously if always assumed IEnumerble and that resulted in a runtime exception when casting later.

Reply:
Added test case with `NameValueCollection`.